lib/sycamore/sycamore/utils/xycut.py

```diff
@@ -0,0 +1,271 @@
+    """Node whose chidren are all Nodes."""
+
+    nodes: list[NodeBase]
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.nodes = []
+
+    def append(self, node: NodeBase) -> "NodeInner":
+        self.nodes.append(node)
+        return self
+
+    def to_str_impl(self, pfx: str, sio: StringIO) -> None:
+        for idx, node in enumerate(self.nodes):
+            node.to_str_impl(f"{pfx}{idx}.", sio)
+
+    def to_text_impl(self, sio: StringIO) -> None:
+        for node in self.nodes:
+            node.to_text_impl(sio)
+
+    def to_elems_impl(self, elems: ElemList) -> None:
+        for node in self.nodes:
+            node.to_elems_impl(elems)
+
+
+class NodeLeaf(NodeBase):
+    """Node whose children are all lists of elements."""
+
+    elists: list[ElemList]
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.elists = [[]]
+
+    def append(self, elem: Element) -> "NodeLeaf":
+        self.elists[-1].append(elem)
+        return self
+
+    def extend(self, elist: ElemList) -> "NodeLeaf":
+        self.elists[-1].extend(elist)
+        return self
+
+    def advance(self) -> "NodeLeaf":
+        self.elists.append([])
+        return self
+
+    def finalize(self) -> "NodeLeaf":
+        if not self.elists[-1]:
+            self.elists.pop()
+        return self
+
+    def cansplit(self) -> bool:
+        try:
+            elem = self.elists[0][0]
+            next = self.elists[1]  # noqa: F841
+            return isinstance(elem, Element)
+        except (IndexError, TypeError):
+            return False
+
+    def to_str_impl(self, pfx: str, sio: StringIO) -> None:
+        for idx, elist in enumerate(self.elists):
+            for i, elem in enumerate(elist):
+                s = elem_to_str(elem)
+                sio.write(f"{pfx}{i}: {s}\n")
+
+    def to_text_impl(self, sio: StringIO) -> None:
+        for elist in self.elists:
+            for elem in elist:
+                if tr := elem.text_representation:
+                    sio.write(tr)
+                    sio.write("\n\n")
+
+    def to_elems_impl(self, elems: ElemList) -> None:
+        for elist in self.elists:
+            elems.extend(elist)
+
+
+###############################################################################
+
+
+def get_bbox(elem: Element) -> tuple[float, float, float, float]:
+    if bbox := elem.data.get("bbox"):
+        return bbox
+    return (1.0, 1.0, 1.0, 1.0)
+
+
+def bbox_to_str(bbox: tuple[float, float, float, float]) -> str:
+    a, b, c, d = [int(1000 * x) for x in bbox]
+    return f"({a:03d},{b:03d})({c:03d},{d:03d})"
+
+
+def elem_to_str(elem: Element) -> str:
+    s = bbox_to_str(get_bbox(elem))
+    if tr := elem.text_representation:
+        nonl = tr.replace("\n", " ")[:32]
+        s += f"[{nonl}]"
+    return s
+
+
+def make_begin_end(elems: ElemList, axis: int) -> BeginEndList:
+    """Returns array of (coord, isopen, elem)"""
+    ary: BeginEndList = []
```

Comment:
ary isn't meaningful, it's just the variable type (which should be lst). 
Use either bel (match the type name); or ret (how it's used); or order (what you used later).